app/src/main/java/org/eyeseetea/malariacare/domain/usecase/pull/DownloadMediaUseCase.java

```diff
@@ -0,0 +1,259 @@
+                                return;
+                            }
+
+                            //Other error
+                            Log.e(TAG, "onCancelled: " + mLastError == null ? "" : mLastError.getMessage());
+                        }
+
+                        @Override
+                        public void onSuccess(HashMap<String, String> syncedFiles) {
+                            int correctSyncedFiles=0;
+
+                            //Try to reuse a local copy if another media references same url
+                            if(syncedFiles.size()>0){
+                                for(Media media:medias){
+                                    String absolutePath = syncedFiles.get(media.getResourceUrl());
+                                    List<Media> allMediaWithSameResource = mMediaRepository.getAllMediaByResourceUid(media.getResourceUrl());
+                                    for(Media localMedia:allMediaWithSameResource) {
+                                        Log.d(TAG, localMedia.toString()+ "\tsaved in " + absolutePath);
+                                        localMedia.setResourcePath(absolutePath);
+                                        mMediaRepository.updateModel(localMedia);
+                                    }
+                                }
+
+                            }
+                            if(correctSyncedFiles>0) {
+                                showMessage(String.format("%d files synced", syncedFiles.size()));
+                            }
+                        }
+
+                        @Override
+                        public void removeIllegalFile(String uid) {
+                            Log.e(TAG, "downloadFile error (file extension not supported)" + uid);
+                            mMediaRepository.deleteModel(mMediaRepository.findByUid(uid));
+                        }
+                    }).execute();
+        }
+    }
+
+    private void initServiceAccountCredential() {
+        try {
+            InputStream privateJsonStream = mContext.getResources().openRawResource(
+                    R.raw.driveserviceprivatekey);
+            serviceCredential = GoogleCredential.fromStream(privateJsonStream).createScoped(
+                    Arrays.asList(SCOPES));
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+
+    }
+
+    private boolean isGooglePlayAppAvailable() {
+        try {
+            PreferencesState.getInstance().getContext().getPackageManager()
+                    .getPackageInfo(GooglePlayServicesUtil.GOOGLE_PLAY_STORE_PACKAGE, 0);
+            return true;
+        } catch (PackageManager.NameNotFoundException e) {
+            return false;
+        }
+    }
+
+    /**
+     * Checks whether the device currently has a network connection.
+     *
+     * @return true if the device has a network connection, false otherwise.
+     */
+    private boolean isDeviceOnline() {
```

Comment:
Download files only when wifi is enabled is a domain requirement but UseCase can not invoke external layer directly, we need to create a port interface inside domain layer as with the interfaces of repositories and then we need to create the adapter that communicates with the external layer as repositories. In this case for example:
Interface: IConnectivityManager inside domain layer and ConnectivityManager inside network layer and this dependency are injected it through the constructor as repositories.
UseCase should not know the Android framework component used for verifying connectivity then if in the future the framework change the use case is not affected